### ISSUE TITLE
Failed to download p2repo when URL includes name and password

### ISSUE THREAD

**Issue by y1yang1** [equodev/equo-ide#156]
```
p2deps {
    into 'compileOnly', {
        p2repo 'http://username:password@example.com/dir'
        install 'com.ibm.icu'
        install 'org.apache.felix.scr'
        install 'org.eclipse.core.commands'
        install 'org.eclipse.core.contenttype'
....
```
Failed to download its content, because we need add authenticator when building okhttp client:
```java
public class P2Client implements AutoCloseable {
	private final Cache metadataResponseCache;
	private final OkHttpClient metadataClient;

	private final P2ClientCache cachingPolicy;
	private final OfflineCache offlineMetadataCache;
	private final JarCache jarCache;
	private final LockFile lock;

	public P2Client() throws IOException {
		this(P2ClientCache.PREFER_OFFLINE);
	}

	public P2Client(P2ClientCache cachingPolicy) throws IOException {
		this.cachingPolicy = cachingPolicy;
		this.jarCache = new JarCache(cachingPolicy);
		long maxSize = 50L * 1024L * 1024L; // 50 MiB
		File p2metadata = CacheLocations.p2metadata();
		if (cachingPolicy.cacheAllowed()) {
			metadataResponseCache = new Cache(new File(p2metadata, "connection"), maxSize);
			metadataClient = new OkHttpClient.Builder().cache(metadataResponseCache).build();
		} else {
			metadataResponseCache = null;
			metadataClient = new OkHttpClient.Builder().build();
		}
		offlineMetadataCache = new OfflineCache(new File(p2metadata, "offline"));
		lock = new LockFile(p2metadata);
	}

```
Something like https://howtoprogram.xyz/2016/11/03/basic-authentication-okhttp-example/

**Comment by nedtwigg** [comment]
PRs welcome :)

**Comment by y1yang1** [comment]
Hi @nedtwigg , maybe something looks like
```diff
diff --git a/solstice/src/main/java/dev/equo/solstice/p2/P2Client.java b/solstice/src/main/java/dev/equo/solstice/p2/P2Client.java
index 59feadfc..c9b7d4d6 100644
--- a/solstice/src/main/java/dev/equo/solstice/p2/P2Client.java
+++ b/solstice/src/main/java/dev/equo/solstice/p2/P2Client.java
@@ -16,6 +16,7 @@ package dev.equo.solstice.p2;
 import java.io.ByteArrayInputStream;
 import java.io.File;
 import java.io.IOException;
+import java.net.URL;
 import java.net.UnknownHostException;
 import java.nio.charset.StandardCharsets;
 import java.util.ArrayDeque;
@@ -28,6 +29,7 @@ import java.util.zip.ZipEntry;
 import java.util.zip.ZipInputStream;
 import javax.xml.parsers.DocumentBuilderFactory;
 import okhttp3.Cache;
+import okhttp3.Credentials;
 import okhttp3.OkHttpClient;
 import okhttp3.Request;
 import org.tukaani.xz.XZInputStream;
@@ -130,6 +132,25 @@ public class P2Client implements AutoCloseable {
 		return true;
 	}
 
+	private static Request buildRequest(String url) {
+		// Check if url contains basic authentication, e.g. http://username:password@example.com/dir
+		String[] info = null;
+		try {
+			URL netUrl = new URL(url);
+			String userInfo = netUrl.getUserInfo();
+			if (userInfo != null) {
+				info = userInfo.split(":");
+			}
+		} catch (Exception ignored) {
+		}
+		Request.Builder builder = new Request.Builder().url(url);
+		if (info != null && info.length == 2) {
+			String credential = Credentials.basic(info[0], info[1]);
+			builder.addHeader("Authorization", credential);
+		}
+		return builder.build();
+	}
+
 	private byte[] getBytes(String url) throws IOException, NotFoundException {
 		if (cachingPolicy.tryOfflineFirst()) {
 			var cached = offlineMetadataCache.get(url);
@@ -141,7 +162,7 @@ public class P2Client implements AutoCloseable {
 			}
 		}
 		if (cachingPolicy.networkAllowed()) {
-			var request = new Request.Builder().url(url).build();
+			var request = buildRequest(url);
 			try (var response = metadataClient.newCall(request).execute()) {
 				if (response.code() == 404) {
 					if (cachingPolicy.cacheAllowed()) {

```

**Comment by nedtwigg** [comment]
Fixed as of `plugin-gradle 1.7.3` and `plugin-maven 1.5.3`.
